drivers/ethernet/eth_xlnx_gem_priv.h

```diff
@@ -0,0 +1,743 @@
+ * (while the buffer descriptors are just an array from the software
+ * point of view, the controller treats them as a ring, in which the
+ * last descriptor's control word has a special last-in-ring bit set).
+ * It contains a pointer to the start of the descriptor array, a
+ * semaphore as a means of preventing concurrent access, a free entry
+ * counter as well as indices used to determine which BD shall be used
+ * or evaluated for the next RX/TX operation.
+ */
+struct eth_xlnx_gem_bdring {
+	/* Concurrent modification protection */
+	struct k_sem		ring_sem;
+	/* Pointer to the first BD in the list */
+	struct eth_xlnx_gem_bd	*first_bd;
+	/* Index of the next BD to be used for TX */
+	uint8_t			next_to_use;
+	/* Index of the next BD to be processed (both RX/TX) */
+	uint8_t			next_to_process;
+	/* Number of currently available BDs in this ring */
+	uint8_t			free_bds;
+};
+
+/**
+ * @brief Constant device configuration data structure.
+ *
+ * This struct contains all device configuration data for a GEM
+ * controller instance which is constant. The data herein is
+ * either acquired from the generated header file based on the
+ * data from Kconfig, or from header file based on the device tree
+ * data. Some of the data contained, in particular data relating
+ * to clock sources, is specific to either the Zynq-7000 or the
+ * UltraScale SoCs, which both contain the GEM.
+ */
+struct eth_xlnx_gem_dev_cfg {
+	uint32_t			base_addr;
+	eth_xlnx_gem_config_irq_t	config_func;
+
+	uint32_t			pll_clock_frequency;
+	uint32_t			clk_ctrl_reg_address;
+	enum eth_xlnx_mdc_clock_divider	mdc_divider;
+
+	enum eth_xlnx_link_speed	max_link_speed;
+	uint8_t				init_phy;
+	uint8_t				phy_mdio_addr_fix;
+	uint8_t				phy_advertise_lower;
+	uint32_t			phy_poll_interval;
+	uint8_t				defer_rxp_to_queue;
+	uint8_t				defer_txd_to_queue;
+
+	enum eth_xlnx_amba_dbus_width	amba_dbus_width;
+	enum eth_xlnx_ahb_burst_length	ahb_burst_length;
+	enum eth_xlnx_hwrx_buffer_size	hw_rx_buffer_size;
+	uint8_t				hw_rx_buffer_offset;
+
+	uint8_t				rxbd_count;
+	uint8_t				txbd_count;
+	uint16_t			rx_buffer_size;
+	uint16_t			tx_buffer_size;
+
+	uint8_t				ignore_ipg_rxer;
+	uint8_t				disable_reject_nsp;
+	uint8_t				enable_ipg_stretch;
+	uint8_t				enable_sgmii_mode;
+	uint8_t				disable_reject_fcs_crc_errors;
+	uint8_t				enable_rx_halfdup_while_tx;
+	uint8_t				enable_rx_chksum_offload;
+	uint8_t				disable_pause_copy;
+	uint8_t				discard_rx_fcs;
+	uint8_t				discard_rx_length_errors;
+	uint8_t				enable_pause;
+	uint8_t				enable_tbi;
+	uint8_t				ext_addr_match;
+	uint8_t				enable_1536_frames;
+	uint8_t				enable_ucast_hash;
+	uint8_t				enable_mcast_hash;
+	uint8_t				disable_bcast;
+	uint8_t				copy_all_frames;
+	uint8_t				discard_non_vlan;
+	uint8_t				enable_fdx;
+	uint8_t				disc_rx_ahb_unavail;
+	uint8_t				enable_tx_chksum_offload;
+	uint8_t				tx_buffer_size_full;
+	uint8_t				enable_ahb_packet_endian_swap;
+	uint8_t				enable_ahb_md_endian_swap;
```

Comment:
Many of these uint8_t look like boolean values, so if we combine them to bool bitfield, then we could save some mem.

Reply:
The entire block below tx_buffer_size, now typed bool instead of uint8_t, is now a bitfield.